File: python/BUILDER_COMPOSITE/BUILDER_COMPOSITE/builder/DirectorReservacion.py
from BUILDER_COMPOSITE.builder.ConstructorReservacionSencilla import ConstructorReservacionSencilla
from BUILDER_COMPOSITE.builder.ConstructorReservacionVIP      import ConstructorReservacionVIP
from BUILDER_COMPOSITE.builder.ConstructorReservacionMultiple import ConstructorReservacionMultiple
from .ConstructorReservacion import ConstructorReservacion

class DirectorReservacion:
    
    def __init__(self) -> None:
        self.constructorReservacion = ConstructorReservacion()
    
    def setConstructor(self, constructorReservacion : ConstructorReservacion):
        self.constructorReservacion = constructorReservacion
    
    def construirReservacion(self, datos):
        if self.constructorReservacion is not None:
            if isinstance(self.constructorReservacion, ConstructorReservacionSencilla):
                self.construirBaseReservacion(datos)
            if isinstance(self.constructorReservacion, ConstructorReservacionVIP):
                self.construirBaseReservacion(datos)
                self.constructorReservacion.construirDatosBeneficios(datos[-1])
            if isinstance(self.constructorReservacion, ConstructorReservacionMultiple):
                self.construirBaseReservacion(datos)
                self.constructorReservacion.construirDatosBeneficios(datos[-1])
                
    def construirBaseReservacion(self, datos):
        self.constructorReservacion.construirDatosCliente(*datos[:5])
        self.constructorReservacion.construirDatosCompra(*datos[5:7])
        self.constructorReservacion.construirDatosFuncion(*datos[7:9],datos[9])

File: php/BUILDER_COMPOSITE/constructores/DirectorReservacion.php
<?php

require_once './constructores/ConstructorReservacionSencilla.php';
require_once './constructores/ConstructorReservacionVIP.php';
require_once './constructores/ConstructorReservacionMultiple.php';

class DirectorReservacion {

    public $constructorReservacion;

    public function setConstructorReservacion($constructorReservacion): void {
        $this->constructorReservacion = $constructorReservacion;
    }

    public function construirReservacion($datos) {
        if (!is_null($this->constructorReservacion)) {
            if ($this->constructorReservacion instanceof ConstructorReservacionSencilla) {
                $this->construirBaseReservacion($datos);
            } elseif ($this->constructorReservacion instanceof ConstructorReservacionVIP) {
                $this->construirBaseReservacion($datos);
                $this->constructorReservacion->construirDatosBeneficio($datos["beneficios"]);
            } elseif ($this->constructorReservacion instanceof ConstructorReservacionMultiple) {
                $this->construirBaseReservacion($datos);
                $this->constructorReservacion->construirDatosBeneficio($datos["beneficios"]);
            }else{
                echo 'Constructor no valido';
            }
        }
    }

    public function construirBaseReservacion($datos) {
        $this->constructorReservacion->construirDatosCliente($datos["nombre"],
                $datos["apellidoP"], $datos["apellidoM"], $datos["edad"], $datos["tipoCliente"]);

        $this->constructorReservacion->construirDatosFuncion($datos["asiento"],
                $datos["hora"], $datos["dia"], $datos["mes"], $datos["anio"], $datos["pelicula"]);

        $this->constructorReservacion->construirDatosCompra($datos["numTarjeta"], $datos["fecha"]);
    }

}
